cpp/src/arrow/compute/exec/aggregate.cc

```diff
@@ -121,27 +127,48 @@ Result<Datum> GroupBy(const std::vector<Datum>& arguments, const std::vector<Dat
   ExecSpanIterator argument_iterator;
 
   ExecBatch args_batch;
-  if (!arguments.empty()) {
-    ARROW_ASSIGN_OR_RAISE(args_batch, ExecBatch::Make(arguments));
+  std::optional<int64_t> inferred_length = ExecBatch::InferLength(arguments);
+  if (!inferred_length.has_value()) {
+    inferred_length = ExecBatch::InferLength(keys);
+  }
+  DCHECK(inferred_length.has_value());
```

Comment:
Are we sure this always succeeds? Otherwise we should return an error to the caller instead of crashing out.

Reply:
Yes. The length should always be inferrable from the `keys`.

Reply:
I'm pushing a commit that turns this into a `ARROW_ASSIGN_OR_RAISE`.